File: src/main/java/com/polygloat/exceptions/ErrorException.java
package com.polygloat.exceptions;

import com.polygloat.constants.Message;
import lombok.Getter;
import org.springframework.http.HttpStatus;

import java.io.Serializable;
import java.util.List;

public abstract class ErrorException extends RuntimeException {
    @Getter
    private final List<Serializable> params;

    @Getter
    private final String code;

    public ErrorException(Message message, List<Serializable> params) {
        this.params = params;
        this.code = message.getCode();
    }

    public ErrorException(Message message) {
        this.code = message.getCode();
        params = null;
    }

    public ErrorResponseBody getErrorResponseBody() {
        return new ErrorResponseBody(this.code, this.params);
    }

    public abstract HttpStatus getHttpStatus();

}

File: src/main/kotlin/io/tolgee/exceptions/ErrorException.kt
package io.tolgee.exceptions

import io.tolgee.constants.Message
import org.springframework.http.HttpStatus
import java.io.Serializable

abstract class ErrorException : RuntimeException {
  val params: List<Serializable>?
  val code: String

  constructor(message: Message, params: List<Serializable>?) : super(message.code) {
    this.params = params
    this.code = message.code
  }

  constructor(message: Message) : this(message, null)

  val errorResponseBody: ErrorResponseBody
    get() = ErrorResponseBody(this.code, params)
  abstract val httpStatus: HttpStatus?
}
